Adding basic coding guidelines/workflows for the NuGet contributor (#3180)
NuGet/NuGet.Client: docs/workflow.md

> # Workflow

Reviewer: ```suggestion
# NuGet.Client Development Workflow
```

Author: Similar to the above, I don't see the point in blasting the repo name everywhere :)

To commit the PR to the repo use the GitHub `Squash and Merge` button. We can't stress this enough. Always use `Squash and Merge` unless an exception is explicitly stated in this document.

- *Do* favor having more than 1 reviewer.
- *Do not* merge too quickly. Wait for at least 24h after the last significant changes before merging unless the change is urgent.
- *Do* address all feedback. Not necessarily by accepting it, but by reaching a resolution with the reviewer. All comments need to be marked as resolved before merging.
- *Do* use GitHub's tooling. Re-request review after all feedback has been addressed.

#### Draft Pull Requests

Draft pull requests are allowed, but should have a clear plan for transition to a review ready pull request.
Draft pull requests will be closed within 30 days.

### Branching strategy

> The active development branch in our repo is `dev`. What we ship comes from the `release-majorminorx` branches.

Reviewer: ```suggestion
The active development branch in our repo is `dev`. What we ship comes from the `release-<major>.<minor>` branches.
```

Author: Fixed.

### Solution and project folder structure and naming

The NuGet.Client repo currently has only one solution file named `NuGet.sln`. We do not want/need to have more than one solution file. 
If deemed necessary by the team, we can consider solution filters at a future point.

> 

Reviewer: Suggestions to add:

- Don't add references to external, non-NuGet packages, unless you have a strong justification for doing so.
- If you need a dependency to a NuGet library, prefer a `ProjectReference` over a `PackageReference` to that NuGet package
- Honor the target frameworks in `common.project.props`

Author: > Don't add references to external, non-NuGet packages, unless you have a strong justification for doing so.

That's a good idea, I think it deserves its own section mentioning source-build etc.

Author: All these are good ideas, but maybe not appropriate for this section. I will add them to a list of future improvements.

> Every project in the NuGet.Client repo should be PackageReference based and if possible (read this as not .NET Framework WPF), an SDK based one. 

Reviewer: ```suggestion
Every project in the NuGet.Client repo should be [PackageReference-based](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files) and, if possible, an [SDK-styled](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2019)  one.
```

> Every project in the NuGet.Client repo should be PackageReference based and if possible (read this as not .NET Framework WPF), an SDK based one. 

Reviewer: Consider using a bulleted list for the guidelines in this section

The production source code is under the `src` folder.
The test source code is under the `test` folder.
The files affecting build are located under the `build` folder.

> The powershell files that are not part of the `Package Manager Console` are located under the `scripts` folder.

Reviewer: ```suggestion
The PowerShell files that are not part of the `Package Manager Console` are located under the `scripts` folder. 
```

Follow the existing pattern for new project files (for example, if NuGet.Commands imports common.props at some point, so should NuGet.MyNewProject). 
Test projects have a different convention for the build customization files they import, so be mindful of that.

### Project naming pattern

The general naming pattern is `NuGet.<area>.<subarea>`.

> 

Reviewer: We should suggest add a description in the project file

Author: Will add a new "how to add a new assembly section"
